addons/web_editor/static/src/scss/web_editor.variables.scss

```diff
@@ -630,6 +630,35 @@ $o-color-extras-nesting-selector: '&, .o_colored_level &';
     @return $color1;
 }
 
+// Given a primary color (and an eventually secondary one),the function returns
+// a sass map defining a basic odoo palette. Eventual overrides (eg. 'menu': 4)
+// are passed by the arbitrary argument.
+@function o-palette($-primary, $-secondary: null, $-overrides...) {
+    $-color-2: $-secondary or increase-contrast(desaturate(mix(complement($-primary), #FFFFFF, 80%), 20%), $-primary);
+
+    $-p: ('o-color-1': $-primary);
+    $-p: map-merge($-p, ('o-color-2': $-color-2));
+    $-p: map-merge($-p, ('o-color-3': adjust-hue(#F5F0F0, hue($-primary))));
+    $-p: map-merge($-p, ('o-color-4': #FFFFFF));
+    $-p: map-merge($-p, ('o-color-5': adjust-hue(#2e1414, hue($-primary))));
```

Comment:
> we removed our automatic palette generation stuff of 12.0 because it was bad. 

I think you misunderstood the objective of this function.
The aim is not to add  a "magic field" that create a palette for the user, or replace the "palette design process" neither...

The function creates a palette with the safest colors possible based on how the color-system treats them, let's say a "blueprint" palette. If the designer wants to create something more "particular" he can override specific values... or not use the function at all.

> If we are to reduce our selection to 20 palettes, why not making them all perfect with readable values?

It's what we're doing :) We tested these palettes internally, like a lot. Colors good 👍

> I am afraid that in one month, there will be 5 more "if" in that function "because stuff".

I see, but I don't think it will never be the case until we'll change how CC use colors.
As you can see there is only one **IF** and it's kind of mandatory.
We have to check if `o-color-5` (that we know) is readable over `o-color-1` (that we don't know).
If false the scheme is adapted, but we don't create new slightly-different colors that are a "just a bit" more readable... we reuse in a different way colors that we know are good for sure.

I may have missed something, but in my test there is only one scenario in which the function fails; if you pass `white` as a primary color (😱), than CC2 title will not be perfectly readable.

Reply:
> The aim is not to add a "magic field" that create a palette for the user

Well, it is a function that returns 5 colors based on one or two, so it is definitely that ;)

> If the designer wants to create something more "particular" he can override specific values... or not use the function at all.

Exactly my point: we now have a function that returns results... that can just be ignored and *will* be ignored by designers ;) I can bet my life on it without being afraid :smile: It feels like this should be a tool for "the palette design process" not something called to generate the actual palettes. I fear we already have a lot of magics going on generating the color combination based on those 5 colors. Having a clear definition of 5 colors seems better to me than calls to a function with one or two colors and sometimes override that is hard to read and understand... but we can keep it, it's the same :shrug: Just another way to write it, and we can always remove it the day there are fifteen `if` inside or the new palettes of version 16.0 do not use the function anymore :smirk: 

> It's what we're doing :) We tested these palettes internally, like a lot. Colors good +1 [...] I may have missed something, but in my test there is only one scenario in which the function fails

I was not questioning the color there, just the way to write them. Again I am extremely surprised you did not want to make more overrides to the result of those function calls (even if the function result can be a good base to start from, I am convinced that when you'll look at the palettes in one month, you'll want to change the color-4 to be 1% more light/dark in palette X or whatever :wink:). But as I said earlier, we can still re-question the function that day.

Reply:
>  but we can keep it, it's the same

Ok. Let's keep it and we'll see later. 
My viewpoint 👇🏻
_____



> Exactly my point: we now have a function that returns results...

But it's not the first one @qsm-odoo , we already have an almost equal [function in the configurator](https://github.com/odoo/odoo/blob/master/addons/website/static/src/components/configurator/configurator.js#L505).
The context is different, agreed, but the concept is not new.

> ...that can just be ignored and will be ignored by designers

Rather than design a palette from scratch, you know have a clear reference starting point (I can't recall how many times have been asked **how much** `color-5` was supposed to be dark).
Thus, if you override the function is not to "fix an issue" (colors are good by default), but it's for aesthetic reasons only (eg. avoid double Italian flags in a default library). Consequentially if the palette has readability issues it's easy to spot the problem (your overrides).

Do we want to avoid overrides at all?
Then we should "enforce colours", in the UI too... but it's not how the system is currently supposed to work.
Do we want to keep overrides?
Great, but we have a couple of flaws to handle, both in the "designers process" and in the UI.
I think this PR is a first step in that direction, more in the next point.

> Having a clear definition of 5 colors seems better to me than calls to a function with one or two colors and sometimes override that is hard to read and understand.

I (partially) agree on that. As you may recall I insisted a lot on a 3 colors-based system (primary, secondary, mid-tone for backgrounds). In my initial version black-_ish_ & white were supposed to be auto-generated and "locked".
It was not an aesthetic choice, but a"design principle" requirement. 
Since we're smarter than others we decided to go for 5  fully customisable colors instead, ok ^^
I wrote guidelines, shared internal slides,  I did 2 OXP talks explaining how to use these colors... but despite any recommendation or "clear definitions":

1. Designers push(_ed_) palettes to their limits
2. Users treat `color-4` & `color-5` as "normal palette colours" (_I don't like this dark `#420000`, let's use rose `#f9c7c7`_ ).

**Point 1**
I should have insisted more on the 3 color-system.
Once failed (😅) I should have stressed more about readability implications of "extremes palettes" with my colleagues , almost 2 yrs ago.
Unfortunately the "extreme palettes" have not been fixed during the last theme ref. neither, making this PR even more urgent IMO.
This new function will ease palette creation process, and will make easier to spot inconsistencies. 

**Point 2**
This is trickier. Despite how many guidelines we can write, for the user the **field's limit** is derived by the **field itself**.
If we allow any color, we can expect any color. We hoped (at lest I hoped) that the CCs examples would have provided enough guidance, but I don't think it's the case anymore.
Moreover, what's the point of letting the user set a potential wrong value hoping that he'll notice and fix it?

I still find very interesting your old idea to "link" `color-4` &  `color-5` with the grayscale field.
It makes even more sense now that the grayscale  field is implemented and works very well, while it's clear that we still struggle handling `color-4` &  `color-5`.
The more I use this color system the more I think that color-4 should be linked to the grayscale as you proposed.

![image](https://user-images.githubusercontent.com/8558351/135757792-a5a86570-c934-42c0-af02-61551b610e80.png)


cc. @pierre-pde

Reply:
> we already have an almost equal function in the configurator.

FWI, I was not too convinced about that one either. But the context is different indeed, my only problem with the function here is that it is *used by designers* to generate *default palettes* which are supposed to be perfect :wink: For me, it should not be part of the *running* code but be an internal tool (but as we are bad at sharing our internal tool with the community, I guess a function in the code is not too bad).

> color system [...] etc

Yep... small problems is that if am not mistaken the grayscale config is stored independently of the color palette (meaning we cannot have a palette linked to a specific default grayscale config). But yes, to think about.

On the other side, I don't think users really struggle about the meaning of the five colors.

1. I don't think many users change the *five* ones
2. For the ones that do it, if they mess up, they can still reset the palette to our well-designed ones
3. ... and more than probably, they would also override all individual CC colors to match what they want.

But agree we should improve anyway for next version of course. If we ever do this https://www.odoo.com/web#id=2463604&action=4043&model=project.task&view_type=form&cids=1&menu_id=4720 to finally get rid of our SCSS config complexity, compatibility problems and theme misconfigurations... :grimacing: